lib/copy/copy.js

```diff
@@ -109,18 +109,33 @@ function copyFileFallback (srcStat, src, dest, opts, cb) {
     const ws = fs.createWriteStream(dest, { mode: srcStat.mode })
     ws.on('error', err => cb(err))
       .on('open', () => rs.pipe(ws))
-      .once('close', () => setDestModeAndTimestamps(srcStat, dest, opts, cb))
+      .once('close', () => setDestModeAndTimestamps(srcStat, src, dest, opts, cb))
   })
 }
 
-function setDestModeAndTimestamps (srcStat, dest, opts, cb) {
-  fs.chmod(dest, srcStat.mode, err => {
+function setDestModeAndTimestamps (srcStat, src, dest, opts, cb) {
+  if (!opts.preserveTimestamps) {
+    return fs.chmod(dest, srcStat.mode, cb)
+  }
+
+  const next = (err) => {
```

Comment:
This part is a bit tricky. (Check the sync version to better understand.)

Basically, the flow is:
1 -> (optional) set writable if not writable
2 --> re-read atime from source
3 ---> set dest atime 
4 ----> set dest mode

Reply:
The complexity stems from the optional step. I'm open to suggestions for making this more readable.